src/dataService/tagDataService.ts

```diff
@@ -182,31 +186,42 @@ export class TagDataService {
    * is actually in the user's list (no foreign key constraint).
    * @param tagInputs
    */
-  public async insertTags(tagInputs: TagCreateInput[]): Promise<void> {
+  public async insertTags(tagInputs: TagSaveAssociation[]): Promise<void> {
     await this.db.transaction(async (trx: Knex.Transaction) => {
       await this.insertTagAndUpdateSavedItem(tagInputs, trx);
       await this.usersMetaService.logTagMutation(new Date(), trx);
     });
   }
 
   private async insertTagAndUpdateSavedItem(
-    tagInputs: TagCreateInput[],
-    trx: Knex.Transaction<any, any[]>
+    tagInputs: TagSaveAssociation[],
+    trx: Knex.Transaction<any, any[]>,
+    timestamp?: Date
   ) {
-    const timestamp = mysqlTimeString(new Date(), config.database.tz);
+    if (tagInputs.length === 0) {
+      return;
+    }
+    const updateTimestamp = mysqlTimeString(
+      timestamp ?? new Date(),
+      config.database.tz
+    );
     const inputData = tagInputs.map((tagInput) => {
       return {
         user_id: parseInt(this.userId),
         item_id: parseInt(tagInput.savedItemId),
         tag: tagInput.name,
-        time_added: timestamp,
-        time_updated: timestamp,
+        time_added: updateTimestamp,
+        time_updated: updateTimestamp,
         api_id: parseInt(this.apiId),
       };
     });
     await trx('item_tags').insert(inputData).onConflict().ignore();
-    const itemIds = tagInputs.map((element) => element.savedItemId);
-    await this.savedItemService.updateListItemMany(itemIds).transacting(trx);
+    const itemIds = uniqueArray(
+      tagInputs.map((element) => element.savedItemId)
+    );
+    await this.savedItemService
```

Comment:
so this essentially just updates the updatedAt time for the savedItem because of the previously success tag association creation? Am I reading that correctly?

Reply:
yes, happy to explain more later... it's how offline-capable (mobile) clients sync right now. If the save timestamp is updated they refresh all the other data associated with it (e.g. tags)

The lack of a more robust sync system is also the reason we can't hard-delete anything right now.